### PR TITLE
[stdlib] More tests for the `bit` module

### PR DESCRIPTION
**Pull request by jiex-liu**
As mentioned in #2862 this PR is used to implement test cases for all the functions that has not been tested originally.

### CODE REVIEW

**Comment by LJ-9801** [comment]
@JoeLoser tests failing at pre-commit? 

```
/__w/modular/modular/Kernels/mojo/stdlib/builtin/_startup.mojo:136:4: error: function instantiation failed
/__w/modular/modular/Kernels/mojo/stdlib/builtin/_startup.mojo:136:4: note:   call expansion failed - no concrete specializations
/__w/modular/modular/Kernels/mojo/stdlib/builtin/_startup.mojo:119:4: note:     function instantiation failed
/__w/modular/modular/Kernels/mojo/stdlib/builtin/_startup.mojo:131:57: note:       call expansion failed - no concrete specializations
/__w/modular/modular/Kernels/mojo/stdlib/builtin/_startup.mojo:91:4: note:         function instantiation failed
/__w/modular/modular/Kernels/mojo/stdlib/builtin/_startup.mojo:107:18: note:           call expansion failed - no concrete specializations
/__w/modular/modular/Kernels/mojo/stdlib/builtin/_startup.mojo:128:8: note:             function instantiation failed
/__w/modular/modular/Kernels/mojo/stdlib/builtin/_startup.mojo:129:22: note:               call expansion failed - no concrete specializations
/home/runner/work/mojo/mojo/stdlib/scripts/check-licenses.mojo:27:5: note:                 function instantiation failed
def main():
    ^
/home/runner/work/mojo/mojo/stdlib/scripts/check-licenses.mojo:38:49: note:                   failed to evaluate 'apply'
        if not file_path.read_text().startswith(LICENSE):
                                                ^
/__w/modular/modular/Kernels/mojo/stdlib/builtin/string.mojo:1653:8: note:                     failed to interpret function @stdlib::builtin::string::String::strip(stdlib::builtin::string::String)
/__w/modular/modular/Kernels/mojo/stdlib/builtin/string.mojo:1660:27: note:                       failed to evaluate call
/__w/modular/modular/Kernels/mojo/stdlib/builtin/string.mojo:1706:8: note:                         failed to interpret function @stdlib::builtin::string::String::lstrip(stdlib::builtin::string::String)
/__w/modular/modular/Kernels/mojo/stdlib/builtin/string.mojo:1714:45: note:                           failed to evaluate call
/__w/modular/modular/Kernels/mojo/stdlib/builtin/string.mojo:585:4: note:                             failed to interpret function @stdlib::builtin::string::_isspace(stdlib::builtin::simd::SIMD[{uint8}, {1}])
/__w/modular/modular/Kernels/mojo/stdlib/utils/static_tuple.mojo:428:44: note:                               failed to interpret operation kgen.struct.gep{index: 0 : index}(#interp.store_to_mem<#kgen.struct<[0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]> : !kgen.struct<(array<256, scalar<ui8>>) memoryOnly>> : !kgen.pointer<struct<(array<256, scalar<ui8>>) memoryOnly>>)
/__w/modular/modular/Kernels/mojo/stdlib/utils/static_tuple.mojo:428:44: note:                                 non-constant inputs
/home/runner/.modular/pkg/packages.modular.com_nightly_mojo/bin/mojo: error: failed to run the pass manager
```

**Review by laszlokindrat — APPROVED** [review]
Thanks for fixing these, it's nice to see improving coverage! My comments are mostly nits, could you please ping me when you addressed them all?

File: stdlib/src/bit/bit.mojo [hunk 1]
```diff
@@ -298,7 +302,7 @@ fn bit_width[
     else:
         var leading_zero_pos = countl_zero(val)
         var leading_zero_neg = countl_zero(bit_not(val))
-        var leading_zero = (val > 0).select(leading_zero_pos, leading_zero_neg)
+        var leading_zero = (val < 0).select(leading_zero_neg, leading_zero_pos)
```
Comment (LJ-9801) [inline stdlib/src/bit/bit.mojo:305]: There was a small bug that cause the `bit_width` of 0 to be 'bitwidthof[type]'. This should fix it.

File: stdlib/test/bit/test_bit.mojo [hunk 2]
```diff
@@ -19,95 +19,378 @@ from bit import (
...
+    alias var1 = SIMD[int8_t, simd_width](-(2**6), 0, -1, 2**6)
+    assert_equal(countr_zero(var1), SIMD[int8_t, simd_width](6, 8, 0, 6))
+
+    alias var3 = SIMD[int16_t, simd_width](-(2**14), 0, -1, 2**14)
+    assert_equal(countr_zero(var3), SIMD[int16_t, simd_width](14, 16, 0, 14))
+
+    alias var5 = SIMD[int32_t, simd_width](-(2**30), 0, -1, 2**30)
+    assert_equal(countr_zero(var5), SIMD[int32_t, simd_width](30, 32, 0, 30))
+
+    # BUG: compiler crash after 2**30 for alias
```
Comment (LJ-9801) [inline stdlib/test/bit/test_bit.mojo:101]: @laszlokindrat somehow any number bigger then 2**30(like 2**31, 2**32) causes a compiler crash for aliasing `SIMD[int64]`

**Comment by LJ-9801** [comment]
@laszlokindrat Any idea why mojo is crashing on `test_reversed.mojo`?

File: stdlib/test/bit/test_bit.mojo (hunk 2 above)
Comment (laszlokindrat) [reply stdlib/test/bit/test_bit.mojo:101]: Could you please file it separately with a minimal reproducer? This probably has something to do with the materialization of `IntLiteral`.

File: stdlib/test/bit/test_bit.mojo [hunk 3]
```diff
@@ -19,95 +19,378 @@ from bit import (
...
+    assert_equal(
+        bit_not(var2), SIMD[int16_t, simd_width](11443, -1, -3001, -8193)
+    )
+
+    alias var3 = SIMD[int32_t, simd_width](-111444, 0, 30000, 2**29)
+    assert_equal(
+        bit_not(var3), SIMD[int32_t, simd_width](111443, -1, -30001, -536870913)
+    )
+
+    # alias var4 = SIMD[int64_t, simd_width](-111444444, 0, 3000000, 2**59)
```
Comment (laszlokindrat) [inline stdlib/test/bit/test_bit.mojo:221]: here and below: Is this another manifestation of the bug you mentioned? If so, can you please link the ticket you open with a #TODO here and everywhere else changes are need to be made?

File: stdlib/src/bit/bit.mojo (hunk 1 above)
Comment (laszlokindrat) [reply stdlib/src/bit/bit.mojo:306]: thanks!

**Comment by laszlokindrat** [comment]
> @laszlokindrat Any idea why mojo is crashing on `test_reversed.mojo`?
> 
> **Update:** Ops it's working now, seems to be very unpredictable tho.

Yeah, I don't think it's related. Could you please rebase after the next nightly?

File: stdlib/test/bit/test_bit.mojo (hunk 3 above)
Comment (LJ-9801) [reply stdlib/test/bit/test_bit.mojo:222]: yep will do!

**Comment by laszlokindrat** [comment]
!sync

**Pull request closed without merging**